src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,466 @@
+        public void defaultVisit(Node node) {
+            // change nothing
+        }
+
+
+        private void visitPreds(Node node, Node[] preds, binding_irnode.ir_opcode opcode) {
+            NodePreds np = new NodePreds(preds, opcode, node.getMode());
+            if (nodeCache.containsKey(np)) {
+                Node n = nodeCache.get(np);
+                if (dominates(n.getBlock().ptr, node.getBlock().ptr)) {
+                    nodeValues.put(node, n);
+                } else {
+                    // replace and update cache
+                    nodeValues.put(n, node);
+                    nodeCache.put(np, node);
+                }
+            } else {
+                nodeCache.put(np, node);
+            }
+        }
+
+
+        @Override
+        public void visit(Add node) {
+            visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Add);
+        }
+
+        @Override
+        public void visit(And node) {
+            visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_And);
+        }
+
+        @Override
+        public void visit(Cmp node) {
+            CmpPreds cmpPreds = new CmpPreds(node.getLeft(), node.getRight(), node.getRelation(), node.getMode());
+            if (cmpCache.containsKey(cmpPreds)) {
+                Node n = cmpCache.get(cmpPreds);
+                if (dominates(n.getBlock().ptr, node.getBlock().ptr)) {
+                    nodeValues.put(node, n);
+                } else {
+                    nodeValues.put(n, node);
+                    cmpCache.put(cmpPreds, node);
+                }
+            } else {
+                cmpCache.put(cmpPreds, node);
+            }
+        }
+
+        @Override
+        public void visit(Cond node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Proj proj) {
+            ProjPreds projPreds = new ProjPreds(proj.getPred(), proj.getNum(), proj.getMode());
+            if (projCache.containsKey(projPreds)) {
+                Node n = projCache.get(projPreds);
+                if (dominates(n.getBlock().ptr, proj.getBlock().ptr)) {
+                    nodeValues.put(proj, n);
+                } else {
+                    nodeValues.put(n, proj);
+                    projCache.put(projPreds, proj);
+                }
+            } else {
+                projCache.put(projPreds, proj);
+            }
+        }
+
+        @Override
+        public void visit(Address addr) {
+            Entity ent = addr.getEntity();
+            if (addressCache.containsKey(ent)) {
+                Node newAddr = addressCache.get(ent);
+                Graph.exchange(addr, newAddr);
+            } else {
+                addressCache.put(ent, addr);
+            }
+        }
+
+        @Override
+        public void visit(Call call) {
+            Attributes a = s.getAttributeAnalysis().getAttributes(Util.getCallee(call));
+            if (a.isPure()) {
```

Comment:
This check is not sufficient. You either need to check `terminates` (or preferably `Util#isPinned`) and `canDeduplicate`.